stumpy/mmotifs.py

```diff
@@ -0,0 +1,177 @@
+        representative to be returned.
+        The first match is always the self-match for each motif.
+
+    max_motifs: int, default 1
+        The maximum number of motifs to return
+
+    atol : float, default 1e-8
+        The absolute tolerance parameter. This value will be added to `max_distance`
+        when comparing distances between subsequences.
+
+    normalize : bool, default True
+        When set to `True`, this z-normalizes subsequences prior to computing distances.
+        Otherwise, this function gets re-routed to its complementary non-normalized
+        equivalent set in the `@core.non_normalized` function decorator.
+
+    p : float, default 2.0
+        The p-norm to apply for computing the Minkowski distance. This parameter is
+        ignored when `normalize == False`.
+
+    Returns
+    -------
+    motif_distances: numpy.ndarray
+        The distances corresponding to a set of subsequence matches for each motif.
+
+    motif_indices: numpy.ndarray
+        The indices corresponding to a set of subsequences matches for each motif.
+
+    motif_subspaces: list
+        A list consisting of arrays that contain the `k`-dimensional
+        subspace for each motif.
+
+    motif_mdls: list
+        A list consisting of arrays that contain the mdl results for
+        finding the dimension of each motif
+    """
+    T = core._preprocess(T)
+    m = T.shape[-1] - P.shape[-1] + 1
+
+    if max_motifs < 1:  # pragma: no cover
+        logger.warning(
+            "The maximum number of motifs, `max_motifs`, "
+            "must be greater than or equal to 1"
+        )
+        logger.warning("`max_motifs` has been set to `1`")
+        max_motifs = 1
+
+    T, M_T, Σ_T = core.preprocess(T, m)
+
+    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
+
+    if max_matches is None:
+        max_matches = np.inf
+
+    if cutoffs is None:
+        P_copy = P.copy().astype(np.float64)
+        P_copy[np.isinf(P_copy)] = np.nan
+        cutoffs = np.nanmax(
+            [
+                np.nanmean(P_copy, axis=1) - 2.0 * np.nanstd(P_copy, axis=1),
+                np.nanmin(P_copy, axis=1),
+            ],
+            axis=0,
+        )
+    if np.isscalar(cutoffs):
+        cutoffs = np.full(P.shape[0], cutoffs)
+
+    motif_distances = []
+    motif_indices = []
+    motif_subspaces = []
+    motif_mdls = []
+
+    candidate_idx = np.argmin(P, axis=1)
+    nn_idx = I[np.arange(len(candidate_idx)), candidate_idx]
+
+    while len(motif_distances) < max_motifs:
+        mdls, subspaces = mdl(T, m, candidate_idx, nn_idx)
+        k = np.argmin(mdls)
+        motif_mdls.append(mdls)
+
+        motif_idx = candidate_idx[k]
+        motif_value = P[k, motif_idx]
+
+        if (
+            motif_value > cutoffs[k]
+            or not np.isfinite(motif_value)
+            or (isinstance(max_distance, float) and motif_value > max_distance)
+        ):
+            break
+
+        motif_subspaces.append(subspaces[k])

```

Comment:
I noticed that we appended the full `mdls` above rather than` mdls[k]`. Would we want to append the full `subspaces` rather than only `subspaces[k]`?

Reply:
I didn't know what you would need all the subspaces for. Do you have an idea? Then I think we would have to return the parameter `k` as well. But if you only pass `subspaces[k]` this is not necessary. Either you don't need `k` at all, or you find it out by iterating over each motif and taking the length of the corresponding line in `motif_subspaces`.

Reply:
Yes, I understand what you mean. The only (mild) concern that I have is that we have one chance to get the API right (i.e., whether to return `subspaces` or `subspaces[k]`). Returning `subspaces` gives the user full flexibility (again, I agree with you that I wouldn't know what else they'd do with it) while `subspaces[k]` have a very specific/limited use.  To some extent, I'm trying to anticipate the future, which is hard! If we choose to return `subspaces[k]` and then people write software that assumes that `subspaces[k]` is returned then we can't later easily change that to return `subspaces` instead since it may break other people's code. 

The only (small) downside to returning `subspaces` instead of `subspaces[k]` is that the user has to take an extra step to extra `k` from `mdls`. But this small annoyance ensures that we haven't "painted ourselves into a corner" and allows this function to be flexible for our users. So, for a small additional step, we give users access to more information.

When it comes to looking at all of the subspaces, I do think that it can be interesting to see if `subspaces[k-1:k+1]` are significantly different. That is, is the set of dimensions at `k` substantially different/similar to its neighboring dimensions. From what I've seen, this is rare but possible. This is interesting because it may suggest that the "important" dimensions have converged to a set.

Reply:
So should I modify `mmotifs()` to return the entire `subspaces`?

Reply:
Yes, that is why I thought that might be pretty confusing for new users.

But wouldn't it be possible to only return `subspaces[k]` and if you would really want to get the whole subspace, couldn't you then later use `stumpy.subspace()`? I think we already have all necessary parameters to do so.

Reply:
> Yes, that is why I thought that might be pretty confusing for new users.

This is an important point that needs to be considered. My opinion (and I am open to being wrong) is that the multi-dimensional concepts are non-trivial and not easily approachable. When that is the case, my goal is to design the functions/APIs so that the tutorials need as few lines of code as possible (aside from the ton of code typically needed to plot stuff) while balancing flexibility. So, I'm already trying to think ahead and I'm inclined to think that the three lines of code above is not too bad and it still allows us maximum flexibility. 

Let me be absolutely clear that this is not a dictatorship and that your feedback/discussion/debate (as the expert user) is extremely relevant here.

Reply:
> motif_distances, motif_indices, motif_mdls, motif_subspaces = mmotifs(...)
> 
> for mdl, subspaces in zip(motif_mdls, motif_subspaces):
>     k = np.argmin(mdl)
>     subspace = subspaces[k]
>     # Do something with `mdl` and `subspace`
    
 But we don't have `mdl` since it is computed in `mmotifs()`. We would have to compute it again outside of the function. That's the reason why I figured out `k` by getting the length of `subspaces[k]` until now. One final thought on my side (don't worry, I'll stop afterwards and will just implement your suggestion 😄): Since the function is called `mmotifs()` aren't we only interested in motif related information? I mean that we only compute the distances and indices for the `motif_idx`, which means that we only consider the time series where the motif with the found dimensionality `k` is present. So if the motif is four dimensional for instance, we are only interested in this four dimensions and don't concider whether we could find the same motif in three or five dimensions (the motif could just be a subset of another motif). Why aren't we then not only interested in finding out WHICH four dimensions are relevant. Why do we also want to find out which dimensions are relevant for the three and four dimensional motif?

Reply:
Ahh, very cool! I really think that this is less confusing. So you let me know when you want me to start implementing the `aamp_mmotifs` function?

Reply:
> So you let me know when you want me to start implementing the aamp_mmotifs function?

@SaVoAMP You have my support! Please go ahead.